llext: add support for init arrays - revised

pillo79 opened the pull request:
This replaces #76724, and includes all feedback received there.

Add support for the `.preinit_array`, `.init_array` and `.fini_array` sections in ELF files. These sections are arrays of function pointers that are filled by the compiler with the addresses of functions, such as C++ constructors, that need to be called at app startup or cleanup by the loader.

This is now achieved separately from `llext_load`, in whatever execution context is deemed secure by the application writer, by either calling directly `llext_bringup`/`llext_teardown` or starting a new user or kernel thread with the `llext_bootstrap` function. The list of function pointers is obtained in all cases via a syscall.  ([link to related doc](https://builds.zephyrproject.io/zephyr/pr/77641/docs/services/llext/load.html#initializing-the-extension)) 

~The PR also adds a `init_fini` test that creates an ELF file with the required sections and verifies the execution order is correct.~ EDIT: Moved to #77941.

~Supporting the `.fini_array` section requires more changes and is not currently implemented.~

pillo79 committed:
llext: add bringup, teardown, and bootstrap APIs

llext_bringup() and llext_teardown() are intended to be used to call the
extension's own initialization and cleanup functions, respectively. They
are meant to be called by the developer after loading an extension and
before unloading it. The list of function pointers to be called is
obtained via the new llext_get_fn_table() syscall, so that they are
compatible with user mode.

llext_bootstrap() is intended to be used as the entry point for a thread
created to run an extension, in either user or kernel contexts. It will
call the extension's own initialization functions and then an additional
entry point in the same context (if desired). The same function can also
be called directly in the main thread, if only initialization is
required.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
doc: llext: add extension bringup and teardown documentation

Add documentation for the new LLEXT APIs that allow to call the
initialization and cleanup functions of an extension.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

mathieuchopstm reviewed (commented):
Looks good overall.

A few thoughts and some minor change suggestions/requests:

pillo79 commented on doc/services/llext/load.rst:
```diff
@@ -24,6 +24,33 @@ successfully, the extension is loaded into memory and is ready to be used.
+   :c:func:`llext_bootstrap` will blindly call all functions mentioned by the
+   sections in the ELF file, as well as the app-provided main entry point, in
+   the context where it is called. This may have security implications if the
+   source of the ELF file is untrusted.
```
There is already #76700 that tracks security-related topics in LLEXT. 
Validation of _the pointers_ is an interesting idea I'll add here (making sure they land in the text section of the extension, at least!), but the issue is with _trusting the code_ in general, and for this, signatures will eventually be required.

pillo79 commented on include/zephyr/llext/llext.h:
```diff
@@ -177,6 +179,48 @@ int llext_load(struct llext_loader *loader, const char *name, struct llext **ext
+ * @param[in] user_data Argument passed to @a entry_fn. Passed as `p3`
+ *                      in @ref k_thread_create.
+ */
+void llext_bootstrap(struct llext *ext, llext_entry_fn_t entry_fn, void *user_data);
```
No. Since `llext_bootstrap` is the "launch vector" for the init functions, it definitely must run in user mode if the app writer wants. The elevated-privileges part is in `llext_get_init_array`.

pillo79 commented on include/zephyr/llext/llext.h:
```diff
@@ -177,6 +179,48 @@ int llext_load(struct llext_loader *loader, const char *name, struct llext **ext
+ * @retval -ENOMEM Init array does not fit in the allocated buffer
+ */
+
+__syscall int llext_get_init_array(struct llext *ext, void *buf, unsigned int size);
```
Not an expert on syscalls myself, but `size_t` appeared to confuse the syscall parser when I tried. Will try again!

pillo79 commented on subsys/llext/llext.c:
```diff
@@ -221,3 +221,37 @@ int llext_call_fn(struct llext *ext, const char *sym_name)
+	int ret = llext_get_init_array(ext, NULL, 0);
+
+	if (ret < 0) {
+		LOG_ERR("Failed to get init array count");
```
Definitely not possible to print the name, this can be run in user context so the `struct llext` content isn't available. I can print the pointer, but would not be the same :slightly_smiling_face:

mathieuchopstm replied:
Fair enough.

pillo79 commented on subsys/llext/llext.c:
```diff
@@ -221,3 +221,37 @@ int llext_call_fn(struct llext *ext, const char *sym_name)
+	}
+
+	/* Start extension main function, if provided */
+	if (entry_fn) {
```
My reasoning was to allow `llext_bootstrap` to be called directly by `k_thread_create`, to minimize boilerplate code. See the [changes in `test_llext_simple.c`](https://github.com/zephyrproject-rtos/zephyr/pull/77641/files#diff-eca0bde030bd631eed59f624aad38146a7b1f54f564a9a77e6b46850453a23ea) and the above linked doc.

teburd replied:
Ahh that makes sense

pillo79 replied:
Absolutely. The LLEXT test suite in this PR already runs `llext_bootstrap` both in user and kernel mode :slightly_smiling_face:

pillo79 commented on subsys/llext/llext_handlers.c:
```diff
@@ -0,0 +1,59 @@
+	if (buf) {
+		char *ptr = buf;
+
+		if (buf_size < preinit_size + init_size) {
+			return -ENOMEM;
+		}
+
+		memcpy(ptr, ext->mem[LLEXT_MEM_PREINIT], preinit_size);
+		memcpy(ptr + preinit_size, ext->mem[LLEXT_MEM_INIT], init_size);
```
`ext->mem_size[]` values must be validated beforehand and are assumed valid everywhere in LLEXT. 
Typically, a buffer of that size has been already allocated in memory (and -4 would cause an error there). There is an issue in the `CONFIG_LLEXT_STORAGE_WRITABLE` case, but that's due to `llext_peek` API being unsafe (no length is given!). I will try to come up with a separate fix for that.

pillo79 committed:
llext: add support for ELF init/fini arrays

Load the .preinit_array, .init_array and .fini_array sections in ELF
files. These sections are arrays of function pointers that are filled by
the compiler with the addresses of functions that need to be called at
startup or termination by the loader, such as C++ constructors and
destructors.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
v2:
- addressed all change suggestions
- added sanity check for returned pointers in `llext_get_init_array`

lyakh commented on subsys/llext/llext.c:
```diff
@@ -221,3 +221,37 @@ int llext_call_fn(struct llext *ext, const char *sym_name)
+
+void llext_bootstrap(struct llext *ext, llext_entry_fn_t entry_fn, void *user_data)
+{
+	int ret = llext_get_init_array(ext, NULL, 0);
```
`ssize_t`?

lyakh commented on subsys/llext/llext.c:
```diff
@@ -221,3 +221,37 @@ int llext_call_fn(struct llext *ext, const char *sym_name)
+
+		ret = llext_get_init_array(ext, &init_fns, sizeof(init_fns));
+		if (ret < 0) {
+			LOG_ERR("Failed to get init array");
```
print error code?

lyakh commented on subsys/llext/llext.c:
```diff
@@ -221,3 +221,37 @@ int llext_call_fn(struct llext *ext, const char *sym_name)
+		typedef void (*elf_init_fn_t)(void);
+
+		int init_fn_count = ret / sizeof(elf_init_fn_t);
+		elf_init_fn_t init_fns[init_fn_count];
```
should we check `ret` for a maximum value before allocating an arbitrary size array on stack?

lyakh commented on subsys/llext/llext_handlers.c:
```diff
@@ -0,0 +1,79 @@
+
+	if (buf) {
+		char *byte_ptr = buf;
+		const char **fn_ptrs = buf;
```
`const void **`? I had to think why `char`, thought there would be some particular purpose in it. Same for `text_start` and `test_end` below. In fact personally I prefer `uint8_t *` over `char *` in such cases for the 2 reasons: (1) these aren't characters or strings, and (2) `uint8_t` has a well defined signedness unlike `char`, but I do realise, that `char` is used a lot in such cases...

lyakh commented on subsys/llext/llext_handlers.c:
```diff
@@ -0,0 +1,79 @@
+
+static int ext_is_valid(struct llext *ext, void *arg)
+{
+	return (ext == arg);
```
superfluous parentheses ;-)

lyakh commented on tests/subsys/llext/simple/src/init_preinit_ext.c:
```diff
@@ -0,0 +1,55 @@
+};
+static const void *const init_fn_ptrs[] __used Z_GENERIC_SECTION(".init_array") = {
+	init_fn
+};
```
can we actually use C++ tests? Might be even nicer to actually have a couple of constructors as a test.

lyakh commented on subsys/llext/llext_handlers.c:
```diff
@@ -0,0 +1,79 @@
+
+	return z_impl_llext_get_init_array(ext, buf, size);
+}
+#include <zephyr/syscalls/llext_get_init_array_mrsh.c>
```
Is this something automatically generated? I don't find this file - neither in the tree nor in this PR? But tests are passing, so I must be missing something

teburd replied:
Yes

pillo79 commented on subsys/llext/llext_load.c:
```diff
@@ -176,6 +178,24 @@ static int llext_find_tables(struct llext_loader *ldr)
+				LOG_ERR("Invalid %s array", sect_type_str);
+				return -ENOEXEC;
+			}
+			break;
```
I started off on the wrong foot, `find_tables` looked nice since it immediately deals with partitions without all the "merge" stuff, but you are right that it isn't an early requirement. Will refactor to use `map_sections` instead.

pillo79 reviewed (commented):
Thanks for the review! Will post an update soon.

pillo79 commented:
Rebased onto current `main` to include #77473, then pushed **v3**:
* moved section identification to `llext_map_sections`
* added support for `.fini_array` as well
* added `llext_bringup` and `llext_teardown` APIs to manually call setup / cleanup functions
* moved tests to separate PR (#77941) to simplify review.

lyakh commented on subsys/llext/llext_load.c:
```diff
@@ -261,14 +289,26 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext)
+				 * referenced at most once in the ELF file.
+				 */
+				LOG_ERR("Region %d redefined", mem_idx);
+				return -ENOEXEC;
```
wouldn't this be caught in line 211?

pillo79 replied:
No, line 211 checks that the _ELF section_ (`ldr->sect_map[shndx]`) was assigned a region before entering `llext_map_sections` - I added that log so the whole function prints a full ELF section table, including reserved sections found by `llext_find_tables`. 
This is instead checking if the _region_ has been already defined (ie. `ldr->sects[mem_idx]` - the array that really should be renamed `ldr->regions`...).

lyakh commented on subsys/llext/llext.c:
```diff
@@ -273,3 +273,66 @@ int llext_call_fn(struct llext *ext, const char *sym_name)
+
+	/* Start extension main function */
+	LOG_DBG("calling entry function %p(%p)", entry_fn, user_data);
+	entry_fn(user_data);
```
you wrote "if desired" in the commit description. Should you check for `entry_fn == NULL` then?

pillo79 replied:
Sorry, commit message is a leftover from the previous iteration, when `llext_bootstrap` was the only way to call init functions. I have since changed code, comments and documentation to make it clear it's not optional anymore; if you only want to init an extension, you now have `llext_bringup`. 
Will fix if I have to do any code change :+1:
